### ISSUE TITLE
Useful hover additions

### ISSUE THREAD

**Issue by arichiardi** [laurentpetit/ccw#765]
I opened this issue in order to discuss which kind of additional hover, together with ```docstring```, I can add to Counterclockwise.

Here are my two ideas:
- Source hover (```clojure.repl/source```) -> I think this can have a huge impact, especially if we can add some sort of "live" editing (changing code in the hover will change the one in the editor).
- Meta hover (```clojure.core/meta```) -> Purely meta information display (I am thinking of inspecting ```clojure.core.memoize``` caches for instance).
- Macro hover (```macroexpand/macroexpand-1/macroexpand-all```) -> For macro expansion, with a switch to select with expansion to show.

This information now can also leverage the [```cider-nrepl```](https://github.com/clojure-emacs/cider-nrepl#supplied-nrepl-middleware) middleware but from if the goal is to still support ```ccw.debug.serverrepl```, everything should be implemented there as well.

**Comment by arichiardi** [comment]
I started to work on the macro expansion hover. Laurent, any suggestion on how to get the top form given an offset? Is there some utility function in ```paredit.clj```?

**Comment by laurentpetit** [comment]
There is a method in Clojure Editor or Source Viewer as far as I remember

2015-05-15 19:20 GMT+02:00 Andrea Richiardi <notifications@github.com>:

> I started to work on the macro expansion hover. Laurent, any suggestion on
> how to get the top form given an offset? Is there some utility function in
> paredit.clj?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laurentpetit/ccw/issues/765#issuecomment-102462086>.
>



-- 
Laurent Petit

**Comment by laurentpetit** [comment]
see `ccw.editors.clojure.editor-support/top-level-code-form`

2015-05-15 20:03 GMT+02:00 Laurent PETIT <laurent.petit@gmail.com>:

> There is a method in Clojure Editor or Source Viewer as far as I remember
>
> 2015-05-15 19:20 GMT+02:00 Andrea Richiardi <notifications@github.com>:
>
>> I started to work on the macro expansion hover. Laurent, any suggestion
>> on how to get the top form given an offset? Is there some utility function
>> in paredit.clj?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/laurentpetit/ccw/issues/765#issuecomment-102462086>.
>>
>
>
>
> --
> Laurent Petit
>



-- 
Laurent Petit

**Comment by laurentpetit** [comment]
There's also `ClojureEditor.getTopLevelSExpression()` and
`ClojureEditor.getCurrentTopLevelSExpression()`.


2015-05-15 20:05 GMT+02:00 Laurent PETIT <laurent.petit@gmail.com>:

> see `ccw.editors.clojure.editor-support/top-level-code-form`
>
> 2015-05-15 20:03 GMT+02:00 Laurent PETIT <laurent.petit@gmail.com>:
>
>> There is a method in Clojure Editor or Source Viewer as far as I remember
>>
>> 2015-05-15 19:20 GMT+02:00 Andrea Richiardi <notifications@github.com>:
>>
>>> I started to work on the macro expansion hover. Laurent, any suggestion
>>> on how to get the top form given an offset? Is there some utility function
>>> in paredit.clj?
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/laurentpetit/ccw/issues/765#issuecomment-102462086>.
>>>
>>
>>
>>
>> --
>> Laurent Petit
>>
>
>
>
> --
> Laurent Petit
>



-- 
Laurent Petit

**Comment by arichiardi** [comment]
Great thank you!

**Comment by asampal** [comment]
@arichiardi have you gotten anywhere with this? It seems quite useful.

**Comment by arichiardi** [comment]
Yes both the docstring and macroexpansion work..try to pull master and please send back some  feedback

**Comment by laurentpetit** [comment]
Oh, we should definitely add something about macroexpansion hover in the
documentation !

2015-07-01 23:24 GMT+02:00 Andrea Richiardi <notifications@github.com>:

> Yes both the docstring and macroexpansion work..try to pull master and
> please send back some feedback
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laurentpetit/ccw/issues/765#issuecomment-117828899>.
>



-- 
Laurent Petit

**Comment by laurentpetit** [comment]
Ok I added documentation for macro expansion in the User Documentation

**Comment by arichiardi** [comment]
Thanks Laurent! I wanted to do it, but busy days these ones..
